File: Node/TypeORM_ORM/src/dal/repositories/orderRepository.ts
import { AppDataSource } from '../../config/data-source';
import { Orders } from '../entities/Orders';
import { TestDatumRepository } from '../../dal/repositories/testDatumRepository';
import { TestData } from '../../dal/entities/TestData';
import { performance } from 'perf_hooks';

//import { GenericRepository } from './genericRepositroy';

export class OrderRepository/* extends GenericRepository<Orders, number>*/ {
  private repo;
  private testDataRepo; 
  constructor() {
    this.repo = AppDataSource.getRepository(Orders);
    this.testDataRepo = new TestDatumRepository();
  }

  async getAll() {
    const start = performance.now();
    const entities = await this.repo.find();
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm Get All Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
    return entities;
  }

  async getById(id: number) {
    const start = performance.now();
    const entity = await this.repo.findOneBy({ orderId: id });
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm Get Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
    return entity;
  }

  async create(data: Partial<Orders>) {
    const start = performance.now();
    const entity = await this.repo.create(data);
    await this.repo.save(entity);
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm Create Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
    return entity;
  }

  async update(id: number, updateData: Partial<Orders>) {
    const start = performance.now();
    const entity = await this.repo.findOneBy({ orderId: id });
    this.repo.merge(entity!, updateData);
    this.repo.save(entity!);
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm Update Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
  }

  async delete(id: number) {
    const start = performance.now();
    await this.repo.delete(id);
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm Update Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
  }

  async allTables(){
    const start = performance.now();
    const result = await AppDataSource.query(`
      SELECT o.*, e.*, c.*, s.*, od.*, p.*, ct.*, sp.*, et.*, t.*, r.*
      FROM Orders o
      JOIN Employees e ON e.EmployeeID = o.EmployeeID
      JOIN Customers c ON c.CustomerID = o.CustomerID
      JOIN Shippers s ON s.ShipperID = o.ShipVia
      JOIN [Order Details] od ON od.OrderID = o.OrderID
      JOIN Products p ON p.ProductID = od.ProductID
      JOIN Categories ct ON ct.CategoryID = p.CategoryID
      JOIN Suppliers sp ON sp.SupplierID = p.SupplierID
      JOIN EmployeeTerritories et ON et.EmployeeID = e.EmployeeID
      JOIN Territories t ON t.TerritoryID = et.TerritoryID
      JOIN Region r ON r.RegionID = t.RegionID`
    );
    const end = performance.now();
    const testDatum = new TestData();
    testDatum.testName = 'Typeorm All Tables Operation';
    testDatum.cpuUsage = `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`;
    testDatum.memoryUsage = `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`;
    testDatum.performance = `${(end - start)} ms`;
    testDatum.language = 'Node.js';
    await this.testDataRepo.create(testDatum);
  }
}

File: Go/GORM_ORM/dal/repositories/orderRepository.go
package repositories

import (
	"GORM_ORM/dal/entities"
	"time"

	"gorm.io/gorm"
)

type OrderRepository interface {
	GetAll() ([]entities.Order, error)
	GetByID(id int) (*entities.Order, error)
	Create(entity *entities.Order) (*entities.Order, error)
	Update(entity *entities.Order) error
	Delete(id int) error
	AllTables() error
}

type orderRepository struct {
	//genericRepo GenericRepository[entities.Order]
	db           *gorm.DB
	testDataRepo TestDataRepository
}

func NewOrderRepository(db *gorm.DB, testDataRepo TestDataRepository) OrderRepository {
	return &orderRepository{
		//genericRepo: NewGenericRepository[entities.Order](db),
		db:           db,
		testDataRepo: testDataRepo,
	}
}

func (r *orderRepository) GetAll() ([]entities.Order, error) {
	var orders []entities.Order
	start := time.Now()
	err := r.db.Find(&orders).Error
	duration := time.Since(start)
	if err != nil {
		return nil, err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM Get All Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return nil, err
	}
	return orders, err
}

func (r *orderRepository) GetByID(id int) (*entities.Order, error) {
	var entity entities.Order
	start := time.Now()
	err := r.db.First(&entity, id).Error
	duration := time.Since(start)
	if err != nil {
		return nil, err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM Get Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return nil, err
	}
	return &entity, nil
}

func (r *orderRepository) Create(entity *entities.Order) (*entities.Order, error) {
	start := time.Now()
	err := r.db.Create(entity).Error
	duration := time.Since(start)
	if err != nil {
		return nil, err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM Create Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return nil, err
	}
	return entity, nil
}

func (r *orderRepository) Update(entity *entities.Order) error {
	start := time.Now()
	err := r.db.Save(entity).Error
	duration := time.Since(start)
	if err != nil {
		return err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM Update Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return err
	}
	return nil
}

func (r *orderRepository) Delete(id int) error {
	var entity entities.Order
	start := time.Now()
	err := r.db.Delete(&entity, id).Error
	duration := time.Since(start)
	if err != nil {
		return err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM Delete Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return err
	}
	return nil
}

func (r *orderRepository) AllTables() error {
	start := time.Now()
	rows, err := r.db.Raw(`
		SELECT o.*, e.*, c.*, s.*, od.*, p.*, ct.*, sp.*, et.*, t.*, r.*
		FROM Orders o
		JOIN Employees e ON e.EmployeeID = o.EmployeeID
		JOIN Customers c ON c.CustomerID = o.CustomerID
		JOIN Shippers s ON s.ShipperID = o.ShipVia
		JOIN [Order Details] od ON od.OrderID = o.OrderID
		JOIN Products p ON p.ProductID = od.ProductID
		JOIN Categories ct ON ct.CategoryID = p.CategoryID
		JOIN Suppliers sp ON sp.SupplierID = p.SupplierID
		JOIN EmployeeTerritories et ON et.EmployeeID = e.EmployeeID
		JOIN Territories t ON t.TerritoryID = et.TerritoryID
		JOIN Region r ON r.RegionID = t.RegionID
	`).Rows()
	duration := time.Since(start)
	_ = rows
	if err != nil {
		return err
	}
	testData := &entities.TestData{
		Language:    stringPtr("Golang"),
		TestName:    stringPtr("GORM All Tables Operation"),
		Performance: stringPtr(duration.String()),
		MemoryUsage: stringPtr( /*formatMemory(memoryUsageMB)*/ "0 MB"),
		CpuUsage:    stringPtr( /*strconv.FormatInt(cpuUsageMs, 10)*/ "0 ms"),
	}
	err = r.testDataRepo.Create(testData)
	if err != nil {
		return err
	}
	return nil
}

func stringPtr(s string) *string {
	return &s
}
